substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/oldobject/JfrOldObjectRepository.java

```diff
@@ -0,0 +1,172 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.oracle.svm.core.jfr.oldobject;
+
+import com.oracle.svm.core.Uninterruptible;
+import com.oracle.svm.core.jfr.JfrBuffer;
+import com.oracle.svm.core.jfr.JfrBufferAccess;
+import com.oracle.svm.core.jfr.JfrBufferType;
+import com.oracle.svm.core.jfr.JfrChunkWriter;
+import com.oracle.svm.core.jfr.JfrNativeEventWriter;
+import com.oracle.svm.core.jfr.JfrNativeEventWriterData;
+import com.oracle.svm.core.jfr.JfrNativeEventWriterDataAccess;
+import com.oracle.svm.core.jfr.JfrRepository;
+import com.oracle.svm.core.jfr.JfrType;
+import com.oracle.svm.core.jfr.SubstrateJVM;
+import com.oracle.svm.core.jfr.traceid.JfrTraceIdEpoch;
+import com.oracle.svm.core.locks.VMMutex;
+import org.graalvm.compiler.word.Word;
+import org.graalvm.nativeimage.Platform;
+import org.graalvm.nativeimage.Platforms;
+import org.graalvm.nativeimage.StackValue;
+import org.graalvm.word.WordFactory;
+
+public final class JfrOldObjectRepository implements JfrRepository {
+    private static final int OBJECT_DESCRIPTION_MAX_SIZE = 100;
+
+    private final VMMutex mutex;
+    private final JfrOldObjectEpochData epochData0;
+    private final JfrOldObjectEpochData epochData1;
+    private long nextId;
+
+    @Platforms(Platform.HOSTED_ONLY.class)
+    public JfrOldObjectRepository() {
+        this.mutex = new VMMutex("jfrOldObjectRepository");
+        this.epochData0 = new JfrOldObjectEpochData();
+        this.epochData1 = new JfrOldObjectEpochData();
+    }
+
+    public void teardown() {
+        epochData0.teardown();
+        epochData1.teardown();
+    }
+
+    @Uninterruptible(reason = "Locking without transition and result is only valid until epoch changes.", callerMustBe = true)
+    public long serializeOldObject(Object obj) {
+        mutex.lockNoTransition();
+        try {
+            JfrOldObjectEpochData epochData = getEpochData(false);
+            if (epochData.buffer.isNull()) {
+                epochData.buffer = JfrBufferAccess.allocate(JfrBufferType.C_HEAP);
+            }
+
+            final Word pointer = Word.objectToUntrackedPointer(obj);
+            JfrNativeEventWriterData data = StackValue.get(JfrNativeEventWriterData.class);
+            JfrNativeEventWriterDataAccess.initialize(data, epochData.buffer);
+            final long objectId = nextId++;
```

Comment:
The initial value is 0 and it's increment before using it, so the first value used would be 1. It would be clearer though if the value was initialized to 1 and incremented after use. I'll amend that.